src/mbgl/geometry/dem_data.cpp

```diff
@@ -3,62 +3,44 @@
 
 namespace mbgl {
 
-DEMData::DEMData(const PremultipliedImage& _image, Tileset::DEMEncoding encoding):
+DEMData::DEMData(const PremultipliedImage& _image, Tileset::DEMEncoding _encoding):
     dim(_image.size.height),
     // extra two pixels per row for border backfilling on either edge
     stride(dim + 2),
+    encoding(_encoding),
     image({ static_cast<uint32_t>(stride), static_cast<uint32_t>(stride) }) {
 
     if (_image.size.height != _image.size.width){
         throw std::runtime_error("raster-dem tiles must be square.");
     }
 
-    auto decodeMapbox = [] (const uint8_t r, const uint8_t g, const uint8_t b){
-        // https://www.mapbox.com/help/access-elevation-data/#mapbox-terrain-rgb
-        return (r * 256 * 256 + g * 256 + b)/10 - 10000;
-    };
-
-    auto decodeTerrarium = [] (const uint8_t r, const uint8_t g, const uint8_t b){
-        // https://aws.amazon.com/public-datasets/terrain/
-        return ((r * 256 + g + b / 256) - 32768);
-    };
-
-    auto decodeRGB = encoding == Tileset::DEMEncoding::Terrarium ? decodeTerrarium : decodeMapbox;
-
-    std::memset(image.data.get(), 0, image.bytes());
-
+    auto dest = reinterpret_cast<uint32_t*>(image.data.get()) + stride + 1;
+    auto source = reinterpret_cast<uint32_t*>(_image.data.get());
     for (int32_t y = 0; y < dim; y++) {
-        for (int32_t x = 0; x < dim; x++) {
-            const int32_t i = y * dim + x;
-            const int32_t j = i * 4;
-            set(x, y, decodeRGB(_image.data[j], _image.data[j+1], _image.data[j+2]));
-        }
+        memcpy(dest, source, dim * 4);
+        dest += stride;
+        source += dim;
     }
     
     // in order to avoid flashing seams between tiles, here we are initially populating a 1px border of
     // pixels around the image with the data of the nearest pixel from the image. this data is eventually
     // replaced when the tile's neighboring tiles are loaded and the accurate data can be backfilled using
     // DEMData#backfillBorder
 
+    auto data = reinterpret_cast<uint32_t*>(image.data.get());
     for (int32_t x = 0; x < dim; x++) {
+        auto rowOffset = stride * (x + 1);
         // left vertical border
-        set(-1, x, get(0, x));
+        data[rowOffset] = data[rowOffset + 1];
```

Comment:
If we have 2px padding on a right side of a stride, would this will overwrite 0th row and create 3px padding on the right?

Reply:
No, starting from row (with index) 1, copies item (row = 1, column = 1) to (row = 1, column = 0)
This is covered well in tests.